### PR TITLE
gh-121368: Fix seq lock memory ordering in _PyType_Lookup

### PR DESCRIPTION
**Pull request by colesbury**
The `_PySeqLock_EndRead` function needs an acquire fence to ensure that
the load of the sequence happens after any loads within the read side
critical section. The missing fence can trigger bugs on macOS arm64.

Additionally, we need a release fence in `_PySeqLock_LockWrite` to
ensure that the sequence update is visible before any modifications to
the cache entry.

Also fix up the return types of a few seqlock functions.

* Issue: gh-121368

### CODE REVIEW

**Comment by ambv** [comment]
After running all repros I had for several hours I confirm this change solves the issue.

**Comment by colesbury** [comment]
Here's the relevant portion of a Rust seqlock implementation for reference:

https://github.com/Amanieu/seqlock/blob/b72653fd9c38141da7e588a97a1b654239f5d4df/src/lib.rs#L138-L141

**Comment by colesbury** [comment]
I've added a model for the C11 model check here: https://github.com/colesbury/c11-model-checker/blob/cpython-models/test/seqlock.c

(Turns out we need the release fence too)

File: Python/lock.c [hunk 1]
```diff
@@ -547,28 +548,31 @@ uint32_t _PySeqLock_BeginRead(_PySeqLock *seqlock)
...
 
-uint32_t _PySeqLock_EndRead(_PySeqLock *seqlock, uint32_t previous)
+int _PySeqLock_EndRead(_PySeqLock *seqlock, uint32_t previous)
 {
-    // Synchronize again and validate that the entry hasn't been updated
-    // while we were readying the values.
-     if (_Py_atomic_load_uint32_acquire(&seqlock->sequence) == previous) {
+    // gh-121368: We need an explicit acquire fence here to ensure that
+    // this load of the sequence number is not reordered before any loads
+    // withing the read lock.
```
Comment (DinoV) [inline Python/lock.c]: withing -> within?

**Review by DinoV — COMMENTED** [review]
LGTM!

**Pull request merged**